Outlier memory usage code adds negative of unsigned number

Visual C++ is issuing this warning:

```
COutliers.cc(931,40): warning C4146: unary minus operator applied to unsigned type, result still unsigned
        recordMemoryUsage(-pointsMemory);
```

The compiler is correct that the value will wrap.  `pointsMemory` is unsigned and `recordMemoryUsage` is a function that takes an unsigned argument.  Therefore the unary minus operator is going to result in a huge number being passed to the function, not a negative number as desired.

However, it probably works as desired because the net effect of adding `pointsMemory` followed by `-pointsMemory` will be to add a total of 2^64 to a 64 bit unsigned number, which will wrap it back to where it started.

This incredibly subtle behaviour is not commented though.  It would be very easy for somebody to assume adding negative numbers to the memory usage works as expected.  I think ideally the way the memory is reduced should be refactored to use subtraction rather than wrapping, and if that's hard then at least the offending code should be commented to explain how it works and why careful thought is required before copying the approach elsewhere.

--- a/include/maths/CDataFrameRegressionModel.h
+++ b/include/maths/CDataFrameRegressionModel.h
@@ -28,7 +28,7 @@ class MATHS_EXPORT CDataFrameRegressionModel {
 public:
     using TDoubleVec = std::vector<double>;
     using TProgressCallback = std::function<void(double)>;
-    using TMemoryUsageCallback = std::function<void(std::uint64_t)>;
+    using TMemoryUsageCallback = std::function<void(std::int64_t)>;
     using TPersistFunc = std::function<void(core::CStatePersistInserter&)>;
     using TTrainingStateCallback = std::function<void(TPersistFunc)>;
 

--- a/include/maths/COutliers.h
+++ b/include/maths/COutliers.h
@@ -41,7 +41,7 @@ using TDouble1VecVec = std::vector<TDouble1Vec>;
 using TDouble1VecVec2Vec = core::CSmallVector<TDouble1VecVec, 2>;
 using TDouble1Vec2Vec = core::CSmallVector<TDouble1Vec, 2>;
 using TProgressCallback = std::function<void(double)>;
-using TMemoryUsageCallback = std::function<void(std::uint64_t)>;
+using TMemoryUsageCallback = std::function<void(std::int64_t)>;
 using TMeanAccumulator = CBasicStatistics::SSampleMean<double>::TAccumulator;
 
 //! Get the distance in the complement space of the projection.

--- a/lib/maths/CBoostedTreeImpl.cc
+++ b/lib/maths/CBoostedTreeImpl.cc
@@ -474,7 +474,7 @@ void CBoostedTreeImpl::train(core::CDataFrame& frame,
 
     m_TrainingProgress.progressCallback(recordProgress);
 
-    std::uint64_t lastMemoryUsage(this->memoryUsage());
+    std::int64_t lastMemoryUsage(this->memoryUsage());
     recordMemoryUsage(lastMemoryUsage);
 
     core::CPackedBitVector allTrainingRowsMask{this->allTrainingRowsMask()};

--- a/lib/maths/COutliers.cc
+++ b/lib/maths/COutliers.cc
@@ -34,6 +34,11 @@ double shift(double score) {
     return std::exp(-2.0) + score;
 }
 
+template<typename T>
+std::int64_t signedMemoryUsage(const T& obj) {
+    return static_cast<std::int64_t>(core::CMemory::dynamicSize(obj));
+}
+
 //! \brief This encapsulates creating a collection of models used for outlier
 //! detection.
 //!
@@ -759,21 +764,21 @@ void CEnsemble<POINT>::CModel::addOutlierScores(const std::vector<POINT>& points
         points_.emplace_back(m_Projection * points[i], index);
     }
 
-    std::int64_t pointsMemory(core::CMemory::dynamicSize(points_));
+    std::int64_t pointsMemory{signedMemoryUsage(points_)};
     recordMemoryUsage(pointsMemory);
-    std::int64_t methodMemoryBeforeRun(core::CMemory::dynamicSize(m_Method));
+    std::int64_t methodMemoryBeforeRun{signedMemoryUsage(m_Method)};
 
     // Run the method.
     TDouble1VecVec2Vec methodScores(m_Method->run(points_, index));
 
-    std::int64_t methodMemoryAfterRun(core::CMemory::dynamicSize(m_Method));
+    std::int64_t methodMemoryAfterRun{signedMemoryUsage(m_Method)};
     recordMemoryUsage(methodMemoryAfterRun - methodMemoryBeforeRun);
 
     // Recover temporary memory.
     m_Method->recoverMemory();
 
-    recordMemoryUsage(std::int64_t(core::CMemory::dynamicSize(m_Method)) - methodMemoryAfterRun);
-    std::int64_t scoresMemoryBeforeAdd(core::CMemory::dynamicSize(scores));
+    recordMemoryUsage(signedMemoryUsage(m_Method) - methodMemoryAfterRun);
+    std::int64_t scoresMemoryBeforeAdd{signedMemoryUsage(scores)};
 
     // Update the scores.
     TDouble1Vec2Vec pointScores(methodScores.size());
@@ -785,8 +790,7 @@ void CEnsemble<POINT>::CModel::addOutlierScores(const std::vector<POINT>& points
         scores[i].add(m_LogScoreMoments, m_RowNormalizedProjection, pointScores);
     }
 
-    recordMemoryUsage(std::int64_t(core::CMemory::dynamicSize(scores)) -
-                      scoresMemoryBeforeAdd - pointsMemory);
+    recordMemoryUsage(signedMemoryUsage(scores) - scoresMemoryBeforeAdd - pointsMemory);
 }
 
 template<typename POINT>
@@ -885,7 +889,7 @@ bool computeOutliersNoPartitions(const COutliers::SComputeParameters& params,
     using TPoint = CMemoryMappedDenseVector<CFloatStorage>;
     using TPointVec = std::vector<TPoint>;
 
-    std::uint64_t frameMemory(frame.memoryUsage());
+    std::int64_t frameMemory{signedMemoryUsage(frame)};
     recordMemoryUsage(frameMemory);
 
     CEnsemble<TPoint>::TScorerVec scores;
@@ -904,7 +908,7 @@ bool computeOutliersNoPartitions(const COutliers::SComputeParameters& params,
         // access and write to each element. Since it does this once per element it
         // is thread safe.
         TPointVec points(frame.numberRows(), TPoint{nullptr, 1});
-        std::uint64_t pointsMemory(core::CMemory::dynamicSize(points));
+        std::int64_t pointsMemory{signedMemoryUsage(points)};
         recordMemoryUsage(pointsMemory);
 
         auto rowsToPoints = [&points](TRowItr beginRows, TRowItr endRows) {
@@ -952,7 +956,7 @@ bool computeOutliersNoPartitions(const COutliers::SComputeParameters& params,
 
     frame.resizeColumns(params.s_NumberThreads,
                         (params.s_ComputeFeatureInfluence ? 2 : 1) * dimension + 1);
-    recordMemoryUsage(frame.memoryUsage() - frameMemory);
+    recordMemoryUsage(signedMemoryUsage(frame) - frameMemory);
 
     bool successful;
     std::tie(std::ignore, successful) = frame.writeColumns(params.s_NumberThreads, writeScores);
@@ -986,7 +990,7 @@ bool computeOutliersPartitioned(const COutliers::SComputeParameters& params,
 
     frame.resizeColumns(params.s_NumberThreads,
                         (params.s_ComputeFeatureInfluence ? 2 : 1) * dimension + 1);
-    recordMemoryUsage(frame.memoryUsage());
+    recordMemoryUsage(signedMemoryUsage(frame));
 
     std::size_t rowsPerPartition{(frame.numberRows() + params.s_NumberPartitions - 1) /
                                  params.s_NumberPartitions};
@@ -997,7 +1001,7 @@ bool computeOutliersPartitioned(const COutliers::SComputeParameters& params,
     // This is presized so that rowsToPoints only needs to access and write to
     // each element. Since it does this once per element it is thread safe.
     TPointVec points(rowsPerPartition, SConstant<TPoint>::get(dimension, 0.0));
-    recordMemoryUsage(core::CMemory::dynamicSize(points));
+    recordMemoryUsage(signedMemoryUsage(points));
 
     for (std::size_t i = 0, beginPartitionRows = 0; i < params.s_NumberPartitions;
          ++i, beginPartitionRows += rowsPerPartition) {
@@ -1046,7 +1050,7 @@ bool computeOutliersPartitioned(const COutliers::SComputeParameters& params,
             return false;
         }
 
-        recordMemoryUsage(-std::int64_t(core::CMemory::dynamicSize(scores)));
+        recordMemoryUsage(-signedMemoryUsage(scores));
     }
 
     return true;

